apps/shared/avifpng.c

```diff
@@ -17,6 +20,119 @@ typedef png_bytep png_iccp_datap;
 typedef png_charp png_iccp_datap;
 #endif
 
+// Converts a hexadecimal string which contains 2-byte character representations of hexadecimal values to raw data.
+// hexString may contain values consisting of [A-F][a-f][0-9] in pairs, e.g., 7af2..., separated by any number of newlines.
+// On success the bytes are filled and AVIF_TRUE is returned.
+// AVIF_FALSE is returned if fewer than expectedLength hexadecimal pairs are converted.
+static avifBool avifHexStringToBytes(const char * hexString, size_t expectedLength, avifRWData * bytes)
+{
+    avifRWDataRealloc(bytes, expectedLength);
+    const char * src = hexString;
+    size_t actualLength = 0;
+    for (; (actualLength < expectedLength) && (*src != '\0'); ++actualLength) {
+        while (*src == '\n') {
+            ++src;
+        }
+        if (!isxdigit(src[0]) || !isxdigit(src[1])) {
```

Comment:
Nit/optional: An earlier version of this pull request does the conversion manually. We can use a variant of that:

```
// Returns the value of the hexadecimal digit. Returns -1 hexDigit is not a hexadecimal digit.
static int avifHexDigitValue(char hexDigit)
{
    if ((hexDigit >= 'A') && (hexDigit <= 'F')) {
        return hexDigit - 'A' + 10;
    } else if ((hexDigit >= 'a') && (hexDigit <= 'f')) {
        return hexDigit - 'a' + 10;
    } else if ((hexDigit >= '0') && (hexDigit <= '9')) {
        return hexDigit - '0';
    } else {
        return -1;
    }
}
```

Reply:
Not sure how this is better than `isxdigit()` + `strtol()`.

Reply:
It is more efficient than `isxdigit()` + `strtol()`, especially if we extend the function to process two hex digits. In any case, it's more important to make sure the parsing code is correct and safe.